File: 01_Array_and_String/C++ Code/06_151. Reverse Words in a String.cpp
// Problem Link -> https://leetcode.com/problems/reverse-words-in-a-string/description/

class Solution {
public:
    string reverseWords(string s) {
        reverse(s.begin(), s.end());
        int right = 0, left = 0;
        int idx = 0;
        int n = s.length();

        while (idx < n) {
            while (idx < n && s[idx] != ' ') {
                s[right++] = s[idx++];
            }

            if (left < right) {
                reverse(s.begin() + left, s.begin() + right);
                s[right] = ' ';
                right++;
                left = right;
            }
            idx++;
        }

        return s.substr(0, right - 1);
    }
};



// Time Complexity -> O(n)
// Space Complexity -> O(1)

File: 01_Array_and_String/Python Code/06_151. Reverse Words in a String.py
# Problem Link -> https://leetcode.com/problems/reverse-words-in-a-string/description/

class Solution:
    def reverseWords(self, s: str) -> str:
        s = list(s[::-1])
        idx, left, right = 0, 0, 0
        n = len(s)

        while idx < n:
            while idx < n and s[idx] != " ":
                s[right] = s[idx]
                idx += 1
                right += 1

            if left < right:
                s[left:right] = reversed(s[left:right])
                if right < n:
                    s[right] = ' '
                right += 1
                left = right

            idx += 1

        return "".join(s[: right - 1])



# Time Complexity -> O(n)
# Space Complexity -> O(1)
